junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TestInstancePreDestroyCallbackTests.java

```diff
@@ -0,0 +1,206 @@
+		);
+		// @formatter:on
+	}
+
+	@Test
+	void classLifecyclePreDestroyCallbacks() {
+		executeTestsForClass(PerClassLifecyclePreDestroyCallbackWithTwoTestMethods.class).testEvents().assertStatistics(
+			stats -> stats.started(2).succeeded(2));
+
+		// @formatter:off
+		assertThat(callSequence).containsExactly(
+			"beforeEachMethod",
+			"test1",
+			"beforeEachMethod",
+			"test2",
+			"fooPreDestroyCallbackTestInstance:PerClassLifecyclePreDestroyCallbackWithTwoTestMethods"
+		);
+		// @formatter:on
+	}
+
+	// -------------------------------------------------------------------
+
+	@ExtendWith(FooInstancePreDestroyCallback.class)
+	static class OuterTestCase extends Named {
+
+		@BeforeEach
+		void beforeOuterMethod() {
+			callSequence.add("beforeOuterMethod");
+		}
+
+		@Test
+		void testOuter() {
+			assertFalse(isDestroyed);
+			callSequence.add("testOuter");
+		}
+
+		@Nested
+		@ExtendWith(BarInstancePreDestroyCallback.class)
+		class InnerTestCase extends Named {
+
+			@BeforeEach
+			void beforeInnerMethod() {
+				assertFalse(isDestroyed);
+				callSequence.add("beforeInnerMethod");
+			}
+
+			@Test
+			void testInner() {
+				callSequence.add("testInner");
+			}
+		}
+	}
+
+	static class TestCaseWithTestSpecificTestInstancePreDestroyCallback extends Named {
+
+		@BeforeEach
+		void beforeEachMethod() {
+			assertFalse(isDestroyed);
+			callSequence.add("beforeEachMethod");
+		}
+
+		@ExtendWith(FooInstancePreDestroyCallback.class)
+		@Test
+		void test() {
+			callSequence.add("test");
+		}
+	}
+
+	@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+	@ExtendWith(FooInstancePreDestroyCallback.class)
+	static class PerClassLifecyclePreDestroyCallbackWithTwoTestMethods extends Named {
+
+		@BeforeEach
+		void beforeEachMethod() {
+			callSequence.add("beforeEachMethod");
+		}
+
+		@Test
+		void test1() {
+			callSequence.add("test1");
+		}
+
+		@Test
+		void test2() {
+			callSequence.add("test2");
+		}
+	}
+
+	static class FooInstancePreDestroyCallback extends AbstractInstancePreDestroyCallback {
+
+		protected FooInstancePreDestroyCallback() {
+			super("foo");
+		}
+	}
+
+	static class BarInstancePreDestroyCallback extends AbstractInstancePreDestroyCallback {
+
+		protected BarInstancePreDestroyCallback() {
+			super("bar");
+		}
+	}
+
+	static abstract class AbstractInstancePreDestroyCallback implements TestInstancePreDestroyCallback {
+
+		private final String name;
+
+		AbstractInstancePreDestroyCallback(String name) {
+			this.name = name;
+		}
+
+		@Override
+		public void preDestroyTestInstance(Object testInstance, ExtensionContext context) {
+			if (testInstance instanceof Named) {
+				((Named) testInstance).setDestroyed();
+				assertTrue(context.getTestInstance().isPresent());
+				assertSame(testInstance, context.getTestInstance().get());
+			}
+			callSequence.add(name + "PreDestroyCallbackTestInstance:" + testInstance.getClass().getSimpleName());
+		}
+	}
+
+	private abstract static class Named {
```

Comment:
I'd rename `Named` into something like `Destroyable` (if this word exists...)

Reply:
yes, thats a good suggestion, renamed!